### PR TITLE
Use provider type name for CodeAction telemetry for code actions...

### LINKED ISSUE #4919: Dont use type names to identify analyzers\fixers for telemetry
We currently create a separate code action for every fixer to identify the fixer for telemetry. We need a different mechanism so that people can just use CodeAction.Create

### PR DESCRIPTION
**Pull request by mavasani**
...created through `CodeAction.Create` factory methods

It only took 7 years, but...
Fixes #4919

Currently, we always use the CodeAction's full type name to generate the telemetry ID to log for applied code actions. This is reasonable for special CodeActions that sub-type the CodeAction type and have custom logic. However, majority of fixers and refactorings do not need special CodeAction sub-type and can/should use the CodeAction.Create factory methods to create the code actions to register. Until now, these providers were forced to create a dummy nested type (generally named MyCodeAction) and use it to allow telemetry to capture the outer type's full type name.

With this change, we now keep track of whether a CodeAction was created with a factory method or not. If it was created with a factory method, then we use the registering fixer/refactoring provider's full type name for telemetry. Otherwise, we use the CodeAction's full type name. Additionally, we also append the EquivalenceKey in the telemetry ID in both the cases, which should allow multiple code actions registered by the same fixer/refactoring to be differentiated.

NOTE: I will create a follow-up PRs in both Roslyn and Roslyn-Analyzers repos to delete all the stub MyCodeAction types once this goes in.

### CODE REVIEW

**Comment by Youssef1313** [comment]
> NOTE: I will create a follow-up PRs in both Roslyn and Roslyn-Analyzers repos to delete all the stub MyCodeAction types once this goes in.

Should this really be done for roslyn-analyzers if Microsoft.CodeAnalysis packages isn't updated to latest?

**Comment by mavasani** [comment]
> Should this really be done for roslyn-analyzers if Microsoft.CodeAnalysis packages isn't updated to latest?

I don’t believe anyone is actively looking at CodeFixer telemetry to be affected if we have a small intermediate period without useful telemetry. So I think we should be fine. Also note that the analyzer package itself doesn’t need to move to reference newer MS.CA packages as there are no public API changes here.

File: src/Workspaces/Core/Portable/CodeActions/CodeAction.cs [hunk 1]
```diff
@@ -369,38 +403,62 @@ public static CodeAction Create(string title, ImmutableArray<CodeAction> nestedA
...
+                CreatedFromFactoryMethod = createdFromFactoryMethod;
             }
 
             public sealed override string Title { get; }
             public sealed override string? EquivalenceKey { get; }
+
+            /// <summary>
+            /// Indicates if this CodeAction was created using one of the 'CodeAction.Create' factory methods.
+            /// </summary>
+            public bool CreatedFromFactoryMethod { get; }
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/CodeActions/CodeAction.cs:427]: Doc why we would need to know this

**Comment by CyrusNajmabadi** [comment]
Will this mean we can get rid of all our 'MyCideAction' subclasses?

**Comment by JoeRobich** [comment]
@mavasani Would you also be able to update the tool that generates a mapping table of CodeAction Provider to TelemetryId? https://github.com/mavasani/roslyn/tree/CodeActionTelemetry/src/Tools/BuildActionTelemetryTable

**Review by jmarolf — APPROVED** [review]

**Comment by jmarolf** [comment]
Thanks @mavasani! No issues with the PR, will just need to remember to keep _both_ kinds of maps in our queries so we don't drop stuff from before this change goes in.

File: src/Workspaces/CoreTest/Shared/Extensions/TelemetryExtensions/TelemetryExtensionTests.cs [hunk 2]
```diff
@@ -5,10 +5,10 @@
 #nullable disable
 
 using System;
-using Microsoft.CodeAnalysis.Editor.Shared.Extensions;
+using Microsoft.CodeAnalysis.Shared.Extensions;
 using Xunit;
 
-namespace Microsoft.CodeAnalysis.Editor.CSharp.UnitTests.Extensions
+namespace Microsoft.CodeAnalysis.UnitTests.Extensions
```
Comment (JoeRobich) [inline src/Workspaces/CoreTest/Shared/Extensions/TelemetryExtensions/TelemetryExtensionTests.cs:11]: This namespace change means the expected telemetryId being tested for has changed to `00000000-0000-0000-c4c5-914100000000`

Comment (JoeRobich) [inline src/Workspaces/CoreTest/Shared/Extensions/TelemetryExtensions/TelemetryExtensionTests.cs:11]: This namespace change means the expected TelemetryId has changed to `00000000-0000-0000-54ad-749900000000`. Also, it seems we are asserting the expected and actual values in the wrong order.

**Comment by mavasani** [comment]
> Would you also be able to update the tool that generates a mapping table of CodeAction Provider to TelemetryId? 

Actually, good observation. I have done partial fixes to that tool, but can get it fully working again only after I am able to cleanup and remove the MyCodeAction nested types. More changes to the tool are coming in the next PR.

**Comment by Youssef1313** [comment]
@mavasani There is an unrelated failure here. Can you restart it so the PR can be merged?

**Pull request merged**